src/jit/codegenxarch.cpp

```diff
@@ -176,23 +176,63 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
 
     // Make sure that EAX is reported as live GC-ref so that any GC that kicks in while
     // executing GS cookie check will not collect the object pointed to by EAX.
-    if (!pushReg && (compiler->info.compRetType == TYP_REF))
-        gcInfo.gcRegGCrefSetCur |= RBM_INTRET;    
+    if (!pushReg)
+    {
+
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+
+        if (varTypeIsStruct(compiler->info.compRetType))
+        {
+            if (compiler->compMethodReturnsMultiRegRetType())
+            {
```

Comment:
Another thing that this routine is not handling is a method returning struct of size > 16 bytes.  In that case that method would be returning RetBufArg (which is a byref) on unix.  Also on windows method would be returning RetBufArg when profiler hooks needs to be generated. I think the logic here should look something like

if (compiler->compMethodReturnsMultiRegRetType())
{
....
}
else if (compiler->compMethodReturnsRetBufAddr())
{
    gcInfo.gcMarkRegPtrVal(REG_INTRET, TYP_BYREF);
}
else
{
    // case of struct returned in a single register and returning scalar types.
    gcInfo.gcMarkRegPtrVal(REG_INTRET, info.compRetNativeType);
}